File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer = 0;
    public int solution(int[] numbers, int target) {
        
        dfs(numbers, 0, target, 0);
        return answer;
    }
    
    public void dfs(int[] numbers, int depth, int target, int result){
        if(depth == numbers.length){
            if(target == result){
                answer++;
            }   
        }
        else{
                dfs(numbers, depth + 1, target, result + numbers[depth]);
                dfs(numbers, depth + 1, target, result - numbers[depth]);
            }
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = 0

    def dfs(depth, result):
        nonlocal answer 
        if depth == len(numbers):
            if result == target:
                answer += 1
        else:
            dfs(depth + 1, result + numbers[depth])
            dfs(depth + 1, result - numbers[depth])

    dfs(0, 0)
    return answer